src/device.rs

```diff
@@ -320,7 +320,36 @@ impl Device {
         //
         // Since we issued a ioctl for getting the destination address, it's safe to assume
         // that if the ioctl was successfull, kernel had set the `ifru_dstaddr` variant.
-        Ok(sockaddr::to_ipv4(unsafe { ifreq.ifr_ifru.ifru_dstaddr }))
+        Ok(unsafe { ifreq.ifr_ifru.ifru_dstaddr }.into())
+    }
+
+    /// Set the hwaddr of the device.
+    /// Will return nix::Error(EOPNOTSUPP) if used on a tun device.
+    pub fn set_hwaddr(&self, hwaddr: [u8; 6]) -> Result<()> {
+        let mut ifreq = self.new_ifreq();
+
+        ifreq.ifr_ifru.ifru_hwaddr = sockaddr::from(hwaddr);
+
+        unsafe {
+            ioctl::siocsifhwaddr(self.inet4_socket.as_raw_fd(), &mut ifreq as *mut bindings::ifreq)?
+        };
+
+        Ok(())
+    }
+
+    /// Get the hwaddr of the device.
+    pub fn get_hwaddr(&self) -> Result<[u8; 6]> {
+        let mut ifreq = self.new_ifreq();
+
+        unsafe {
+            ioctl::siocgifhwaddr(self.inet4_socket.as_raw_fd(), &mut ifreq as *mut bindings::ifreq)?
+        };
+
+        // Safety:
+        //
+        // Since we issued a ioctl for getting the hardware address, it's safe to assume
+        // that if the ioctl was successfull, kernel had set the `ifru_hwaddr` variant.
+        Ok(unsafe { ifreq.ifr_ifru.ifru_hwaddr }.into())
     }
```

Comment:
`Device` is a struct that holds the functions that are shared between a `Tun` and a `Tap` device. As you've commented yourself, these two functions cannot be called from a `Tun` and are exclusive to a `Tap` device. So we should move these functions to `impl` block of the `Tap` struct.

The problem is that these functions won't be exposed from the `AsyncTap` struct since `AsyncTap` wraps around `AsyncDevice` which itself wraps around `AsyncFd<Device>`. So I think to properly support these functions we need a new hierarchical structure that automatically exposes `Tap` functions from `AsyncTap`.

This way we make it impossible to call these functions from any `Tun` device and also automatically expose it to higher level structs that should rely on `Tap`.

Reply:
Introducing a trait for a one time use case is not ideal in my opinion. I was thinking something along the lines of:
```rust
struct Device;

struct Tap(Device);

impl AsRef<Device> for Tap {
    fn as_ref(&self) -> &Device {
        &self.0
    }
}

struct AsyncDevice<T>(T) where T: AsRef<Device>;

struct AsyncTap(AsyncDevice<Tap>);

impl Deref for AsyncTap {
    type Target = Tap;

    fn deref(&self) -> &Self::Target {
        self.get_ref()
    }
}
```
So we can access `set_hw` and `get_hw` using `async_tap.get_hw()` and access `Device` functions using `async_tap.as_ref()`

Reply:
Why don't we just resolve this through typestate, i.e. create a blocking device and a non-blocking asyncdevice, which take a generic parameter, this would IMO safe us a lot of intransparent derefing.